rust/operator/src/lib.rs

```diff
@@ -0,0 +1,674 @@
+    /// Retrieve nodes that fit selectors and store them for later processing:
+    /// SupersetRole (we only have 'node') -> role group -> list of nodes.
+    async fn init_reconcile_state(
+        &self,
+        context: ReconciliationContext<Self::Item>,
+    ) -> Result<Self::State, Self::Error> {
+        let existing_pods = context
+            .list_owned(build_common_labels_for_all_managed_resources(
+                APP_NAME,
+                &context.resource.name(),
+            ))
+            .await?;
+        trace!(
+            "{}: Found [{}] pods",
+            context.log_name(),
+            existing_pods.len()
+        );
+
+        let superset_spec: SupersetClusterSpec = context.resource.spec.clone();
+
+        let mut eligible_nodes = HashMap::new();
+
+        eligible_nodes.insert(
+            SupersetRole::Node.to_string(),
+            find_nodes_that_fit_selectors(&context.client, None, &superset_spec.nodes).await?,
+        );
+
+        let mut roles = HashMap::new();
+        roles.insert(
+            SupersetRole::Node.to_string(),
+            (
+                vec![PropertyNameKind::Env],
+                context.resource.spec.nodes.clone().into(),
+            ),
+        );
+
+        let role_config = transform_all_roles_to_config(&context.resource, roles);
+        let validated_role_config = validate_all_roles_and_groups_config(
+            &context.resource.spec.version.to_string(),
+            &role_config,
+            &self.config,
+            false,
+            false,
+        )?;
+
+        Ok(SupersetState {
+            context,
+            existing_pods,
+            eligible_nodes,
+            validated_role_config,
+        })
+    }
+}
+
+/// Return a map where the key corresponds to the role_group (e.g. "default", "10core10Gb") and
+/// a tuple of a vector of nodes that fit the role_groups selector description, and the role_groups
+/// "replicas" field for scheduling missing pods or removing excess pods.
+pub async fn find_nodes_that_fit_selectors<T>(
+    client: &Client,
+    namespace: Option<String>,
+    role: &Role<T>,
+) -> OperatorResult<HashMap<String, EligibleNodesAndReplicas>>
+where
+    T: Serialize,
+{
+    let mut found_nodes = HashMap::new();
+    for (group_name, role_group) in &role.role_groups {
+        let selector = role_group.selector.to_owned().unwrap_or_default(); // krustlet::add_stackable_selector(role_group.selector.as_ref());
+        let nodes = client
+            .list_with_label_selector(namespace.as_deref(), &selector)
+            .await?;
+        debug!(
+            "Found [{}] nodes for role group [{}]: [{:?}]",
+            nodes.len(),
+            group_name,
+            nodes
+        );
+        found_nodes.insert(
+            group_name.clone(),
+            EligibleNodesAndReplicas {
+                nodes,
+                replicas: role_group.replicas,
+            },
+        );
+    }
+    Ok(found_nodes)
```

Comment:
Can be removed once https://github.com/stackabletech/operator-rs/pull/255 is merged

Reply:
removed